inc/functions-related.php

```diff
@@ -0,0 +1,395 @@
+	if ( $url ) {
+		$html_content = blog_get_html_content( $url );
+		if ( $html_content ) {
+			$data = array();
+			$html = new DOMDocument();
+			@$html->loadHTML( $html_content );
+			$data['title']          = '';
+			$data['description']    = '';
+			$data['img']            = get_stylesheet_directory_uri().'/img/card-thumb.jpg';
+			$data['og-img'][0]      = '';
+			$data['twitter-img']    = '';
+			$data['url']            = $url;
+			$data['type']           = 'National Archives Blog';
+			$i                      = 0;
+			foreach ( $html->getElementsByTagName( 'meta' ) as $meta ) {
+				if ( $meta->getAttribute( 'property' ) == 'og:title' ) {
+					$title = $meta->getAttribute( 'content' );
+					$title = str_replace('| The National Archives blog', '', $title);
+					$title = str_replace('| The National Archives', '', $title);
+					$data['title'] = $title;
+				}
+				if ( $meta->getAttribute( 'property' ) == 'og:description' ) {
+					$data['description'] = $meta->getAttribute( 'content' );
+				}
+				if ( $meta->getAttribute( 'name' ) == 'twitter:image' ) {
+					$data['twitter-img'] = $meta->getAttribute( 'content' );
+				}
+				if ( $meta->getAttribute( 'property' ) == 'og:image' ) {
+					$data['og-img'][ $i ] = $meta->getAttribute( 'content' );
+					$i ++;
+				}
+			}
+			if ( strpos( $url, 'research-guides' ) !== false ) {
+				$data['type'] = 'Research guide';
+			}
+			if ( strpos( $url, 'media.nationalarchives' ) !== false ) {
+				$data['type'] = 'Archives Media Player';
+			}
+			if ( $data['og-img'][0] ) {
+				$data['img'] = $data['og-img'][0];
+			} elseif ( $data['twitter-img'] ) {
+				$data['img'] = $data['twitter-img'];
+			}
+			return $data;
+		}
+	}
+	return false;
+}
+
+function r_gtm( $title, $id, $format, $i ) {
+	$html = 'data-gtm-name="%s" data-gtm-id="%s" data-gtm-position="%s" data-gtm-creative="%s"';
+
+	return sprintf( $html, $title, $id.'-'.$i, 'body-content-'.$i, 'blog-'.$format );
+}
+
+function r_html( $id, $url, $img, $title, $type, $format='cards', $i ) {
+
+	$gtm = r_gtm( $title, $id, $format, $i );
+
+	if ( $format == 'cards' ) {
+
+		$html = '
+<div class="col-md-4 related-post parent-post-%s">
+	<a href="%s" %s class="homepage-card"><div class="related-post-thumb" style="background-image: url(%s)"></div></a>
+	<p><small>%s</small></p>
+	<a href="%s" %s class="homepage-card"><h4>%s</h4></a>
+</div>';
+
+		return sprintf( $html, $id, $url, $gtm, $img, $type, $url, $gtm, $title );
+
+	} else {
+
+		$html = '
+<li class="related-post parent-post-%s">
+	<a href="%s" %s class="homepage-card"><h5>%s</h5></a>
+	<p><small>%s</small></p>
+</li>';
+
+		return sprintf( $html, $id, $url, $gtm, $title, $type );
+
+	}
+}
+
+function get_links_from_content( $content ) {
+
+	$content_links = array();
+	$dom           = new domDocument;
+	@$dom->loadHTML( $content );
+	$dom->preserveWhiteSpace = false;
+	$links                   = $dom->getElementsByTagName( 'a' );
+
+	if ( $links ) {
```

Comment:
There is quite a lot of use of `strpos` and `preg_grep` here which made it quite difficult for me to follow. I'm reading the overall logic as: if the `href` contains 'x, y or z', and the `href` isn't already in `$content_links` then add it. If that's the case I wonder if something like this might be a bit easier to read. 


```php
if ($links) {
    foreach ($links as $link) {

        $href = $link->getAttribute('href');

        if (strpos($href, '.jpg')) {
            break;
        }

        $link_subscrings_of_interest = ['research-guides', 'media.nationalarchives', 'blog.nationalarchives'];

        $substring_in_href = false;

        foreach ($link_subscrings_of_interest as $needle) {
            if (strpos($href, $needle) !== false) {
                $substring_in_href = true;
            }
        }

        if (!$substring_in_href || in_array($href, $content_links)) {
            break;
        }

        array_push($content_links, $href);

    }
}
return $content_links;
```
I'm sure it works fine as it is (and - since I can't currently run WordPress - I haven't had a chance to test the approach I'm proposing) but I wonder if this might be a bit simpler.

Reply:
This looks good. Thanks. Although, it's missing a condition. The resulting array should only have one link of each type. I like what you have done. I'll try work it into the final app if the prototype goes forward.